### ISSUE TITLE
Irrelevant projections are inconsistent

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#543]
Unfortunatly, despite their usefulness, irrelevant record projections are unsound.  Since I can project from the Squash type in irrelevant contexts, I can prove true=false in irrelevant contexts.  But an absurdity in an irrelevant context is as bad as on the top context, whereto it can be lifted.

``` agda
record Squash {ℓ}(A : Set ℓ) : Set ℓ where
  constructor squash
  field
    .unsquash : A
open Squash

-- ok:
sqT≡sqF : squash true ≡ squash false
sqT≡sqF = refl

-- this should not be provable!!
.irrT≡F : true ≡ false
irrT≡F = subst (λ s → unsquash (squash true) ≡ unsquash s) sqT≡sqF refl

-- the rest is easy
T≠F : true ≡ false → ⊥
T≠F p = subst T p tt

.irr⊥ : ⊥
irr⊥ = T≠F irrT≡F     

rel⊥ : .⊥ → ⊥
rel⊥ ()

absurd : ⊥
absurd = rel⊥ irr⊥
```

Original issue reported on code.google.com by `andreas....@gmail.com` on 31 Dec 2011 at 12:21

Attachments:
- [IrrProj.agda](https://storage.googleapis.com/google-code-attachments/agda/issue-543/comment-0/IrrProj.agda)

**Comment by gallais** [comment]
Has this been fixed even when irrelevant projections are allowed?
(In which case: should we go back to considering them safe by default?)

```agda
{-# OPTIONS --irrelevant-projections #-}

open import Data.Bool.Base
open import Data.Empty
open import Data.Unit
open import Relation.Binary.PropositionalEquality.Core

record Squash {ℓ}(A : Set ℓ) : Set ℓ where
  constructor squash
  field
    .unsquash : A
open Squash

-- ok:
sqT≡sqF : squash true ≡ squash false
sqT≡sqF = refl

-- this should not be provable!!
.irrT≡F : true ≡ false
irrT≡F = subst (λ s → unsquash (squash true) ≡ unsquash s) sqT≡sqF refl
```

fails with version 2.6.0:

```
 .(true) != true of type Bool
when checking that the inferred type of an application
  .(true) ≡ unsquash _y_22
matches the expected type
  true ≡ false
```

I looked into this because I was forced to move from projections on the RHS
to patterns on the LHS to port agdarsec to 2.6.0.

**Comment by jespercockx** [comment]
The (known) inconsistency with irrelevant projections was fixed in #2170. As far as I understand irrelevant projections are now consistent but non-conservative (they imply some form of non-constructive choice). I don't know whether it's best to enable them by default but provide a flag to disable them (as for `--without-K`) or to keep the current situation where they are hidden under a flag by default.

(Gratuitous self-promotion: you can avoid all of these problems by using `Prop` instead of irrelevance).
